pennylane/gradients/pulse_gradient.py

```diff
@@ -0,0 +1,532 @@
+        list[QuantumTape]: new tapes with replaced operation, one tape per group of operations in
+        ``split_evolve_ops``.
+    """
+    ops_pre = tape.operations[:op_idx]
+    ops_post = tape.operations[op_idx + 1 :]
+    return [
+        qml.tape.QuantumScript(ops_pre + split + ops_post, tape.measurements)
+        for split in split_evolve_ops
+    ]
+
+
+def _parshift_and_integrate(results, cjacs, int_prefactor, single_measure, shot_vector):
+    """Apply the parameter-shift rule post-processing to tape results and contract
+    with classical Jacobians, effectively evaluating the numerical integral of the stochastic
+    parameter-shift rule.
+    This is a helper method for the new return type system.
+
+    Args:
+        results (list): Tape evaluation results, corresponding to the modified quantum
+            circuit result when using the applicable parameter shifts and the sample splitting
+            times. Results should be ordered such that the different shifted circuits for a given
+            splitting time are grouped together
+        cjacs (tensor_like): classical Jacobian evaluated at the splitting times
+        int_prefactor (float): prefactor of the numerical integration, corresponding to the size
+            of the time range divided by the number of splitting time samples
+        single_measure (bool):
+        shot_vector (bool):
+    Returns:
+        tensor_like or tuple[tensor_like] or tuple[tuple[tensor_like]]: Gradient entry
+    """
+
+    def _diff_and_contract(res_list, cjacs, int_prefactor):
+        diff = qml.math.stack(res_list)[::2] - qml.math.stack(res_list)[1::2]
+        return qml.math.tensordot(diff, cjacs, axes=[[0], [0]]) * int_prefactor
+
+    # If multiple measure xor shot_vector: One axis to pull out of the shift rule and integration
+    if not single_measure + shot_vector == 1:
+        return tuple(_diff_and_contract(r, cjacs, int_prefactor) for r in zip(*results))
+    if single_measure:
+        # Single measurement without shot vector
+        return _diff_and_contract(results, cjacs, int_prefactor)
+
+    # Multiple measurements with shot vector
+    return tuple(
+        tuple(_diff_and_contract(_r, cjacs, int_prefactor) for _r in zip(*r)) for r in zip(*results)
+    )
+
+
+def _stoch_pulse_grad(tape, argnum=None, num_samples=1, sampler_seed=None, shots=None):
+    r"""Compute the gradient of a quantum circuit composed of pulse sequences by applying the
+    stochastic parameter shift rule.
+
+    Args:
+        tape (pennylane.QNode or .QuantumTape): quantum tape or QNode to differentiate
+        argnum (int or list[int] or None): Trainable tape parameter indices to differentiate
+            with respect to. If not provided, the derivatives with respect to all
+            trainable parameters are returned.
+        num_samples (int): number of time samples to use in the stochastic parameter-shift
+            rule underlying the differentiation; also see details
```

Comment:
```suggestion
        num_samples (int): number of time samples to use for the Monte-Carlo sampling in the stochastic parameter-shift
            rule underlying the differentiation; also see details
```

Reply:
If we provide the basic formula in the description above it could also make sense to call this `num_taus` instead? (not sure, vague suggestion)

Reply:
Something more descriptive would help, just `num_samples` could be for example confused with meaning `shots`. I think `num_splitting_times` is not bad, alternatives: just `splitting_times` or `sampling_times` or `num_sampling_times`?

Reply:
Actually. `num_samples` was an outdated name, for SPSA it's `num_directions` now :see_no_evil: 
I think I'd go for `num_split_times`, keeping it a bit shorter.